### ISSUE TITLE
ToneTag - logistics BAP - compliance check

### ISSUE THREAD

**Issue by abhinavv245** [ONDC-Official/v1.2.0-logs#200]
## All Flows

**search**

- Holiday date should not be past dated

**/init**

- /init/0/context/city must be equal to constant (std:040)
- billing details will be of the logistics buyer (not the end buyer)
- /init/0/message/order/billing/tax_number should be valid
- /init/0/message/order/payment/type should be same as in /search - "ON-FULFILLMENT"
- real values should be provided instead of dummy values (XXXXXXXXXXXXXXX)
- settlement details are not required in /init as they will be provided by seller NP (optional) in case payment is collected by BAP

**/confirm**

- /confirm/0/message/order/fulfillments/0/start/instructions must have required property 'code' (invalid schema error, NACK by LSP)
- /confirm/0/message/order/fulfillments/0/end/instructions must have required property 'code'
- /confirm/0/message/order/payment @ondc/org/collection_amount is required only for payment/type 'ON-FULFILLMENT'
- /confirm/0/message/order/payment/@ondc/org/collection_amount mismatches in /payment from /search
- quote object mismatches with the one provided in /on_init (tax breakup is missing) (should be NACK'd by LSP)
- Average Pickup Time mismatches from the one provided in /on_search (should be NACK'd by LSP)

**/update**

- deferred RTS (ready to ship) call is missing (should be handled in one of the flows)

**/on_status**

- tracking should be enabled (true) for hyperlocal (Immediate Delivery)
-  All fulfillment state on_status calls should be provided (Agent-assigned,Order-picked-up,Out-for-delivery,Order-delivered,RTO-Delivered/Disposed)

**/track**

- /track/0/message/order_id must be equal to constant (d8faa255-7fd7-4593-8acb-e3ef91d843ea)

**/on_track**

- /on_track/0/message/tracking/status should be 'active' if tracking is enabled

**/cancel**

- context/action is incorrect
- invalid schema inside messge (order_id is missing, order, order/state are not required)

**/on_cancel**

- in /items rto item/time should be same as provided in /on_search
- average pickup time cannot change with the one provided in /on_search 
- payment/status is not required

**Comment by jagadeeswar577** [comment]
@abhinavv245 

Our merchant's stores are located at Bengaluru and our flows (Retail and Logistic) are tested for the same location.

So the below feedback is looking not relevant in this case.

 **/init/0/context/city must be equal to constant (std:040)** 

And all the above feedbacks are  looking like for logistic payloads , is that understanding correct?

Thanks

**Comment by abhinavv245** [comment]
@jagadeeswar577 the mentioned city code in context should match the city for fulfillment start location which in this case is Telangana as per the start/location/area_code: "500024 (Telangana) provided in Flow 1(Delivery). Kindly reverify all the logs.

Yes, this issue board is specifically for logistics logs.

**Comment by jagadeeswar577** [comment]
![Screenshot 2023-12-18 at 3 34 40 PM](https://github.com/ONDC-Official/v1.2.0-logs/assets/6917215/2d23dfd8-dd40-4117-aa2f-243df658f6a1)

As you can see in the attached screenshot, we have run the utility tool against all 3 flows and it reported zero issue.

Even the only one that we see "Message Id cannot be same for different sets of APIs" is appearing even though we are using different Message ID per API combo.

Can you let us know from where the above mentioned issues are raised? Is it from the same tool? If yes, why we didn't see them in our local machine??

@abhinavv245 

Thanks

**Comment by abhinavv245** [comment]
Hi @jagadeeswar577 Did you use the updated version of the tool (updated on 11 Dec, 2023)

**Comment by jagadeeswar577** [comment]
@abhinavv245 No, we downloaded that utility 7th December. If

**Comment by jagadeeswar577** [comment]
@abhinavv245 

I have the latest log-verification-utility running in my local machine now.

And when I run the Cancel flow logs against it like below

curl -X POST -d "logPath=/Users/jagadeeswardevalla/Downloads/Logistics/Flow-3-Cancel" http://localhost:5000/validate/logistics

**{"logReport":{}}**

I see no issues reported here.

**But you have listed below issues. So how we can check and fix it if we don't see any issues here in local??**

![Screenshot 2023-12-18 at 5 01 07 PM](https://github.com/ONDC-Official/v1.2.0-logs/assets/6917215/b51f1a4e-76fa-40f4-b0be-be1085b520c4)

**Comment by abhinavv245** [comment]
@jagadeeswar577 please make sure the correct path is given, when I running the utility using the cURL or Postman, it is generating the complete log report.

**Comment by abhinavv245** [comment]
## Flow 1 (Order delivered)

**/confirm**

- order/id should be unique and not be equal to transaction_id: \"e2c3ec62-0f63-4a15-a120-61fba90f1b63\"
- tracking is not required in /fulfillments (as it is enabled/disabled by LSP)
-  logistics buyer NP must accept LSP terms (in order/tags). If not accepted, LSP can NACK /confirm with error code 65002, incorrect placement of tags inside linked order

**/update**

- time is not required in /items (time/duration mismatches from the one provided by LSP in /on_search)
- message/order does not require state, provider, quote, billing, payment, created_at (also, quote price does not match the total breakup price)

**/on_update**

- why is fulfillment tracking disabled
- payment/status has changed to 'NOT-PAID' from 'PAID' in /on_confirm (payment/status is required in /on_status only)

## Flow 2 (RTO)

- on_cancel file is missing
- all the on_status calls have the same timestamp
- fulfillment of type 'RTO' is missing 
- RTO charges were not added after RTO was triggered

## Flow 3 - (Cancel)

**/on_cancel**

- order/updated_at is not updated correctly

**/on_status**

- cancellation is required once the order is cancelled
- RTO item will also be added once the order is cancelled and RTO is initiated

@jagadeeswar577

**Comment by jagadeeswar577** [comment]
Hi @abhinavv245 

For **Flow 2 (RTO)** , 

when customer wants to return the item(s) back to merchant due to some reason, we are creating a new service request to the logistic delivery to pick up from customer location(start) to the merchant location(end).

So why here **on_cancel file** comes ?

For **Flow 3 - (Cancel)**,

We are cancelling the Logistic Service request as buyer cancel the order before the item is picked by logistic partner.

So why here **RTO item will also be added once the order is cancelled and RTO is initiated** comes?

**Comment by abhinavv245** [comment]
@jagadeeswar577 

Flow 2 RTO (return to origin) needs to be tested, it means the buyer refused to accept the order or was not available, so the LSP cancelled the order (on_cancel) and initiated the RTO (backward shipment is added). It seems you are confusing it with another use case (return) which will be a new order altogether. 

Flow 3 (Cancel) - As per the logs submitted, order was picked up at "2023-12-19T11:10:00.073Z" and was cancelled at "2023-12-19T11:10:05.373Z" , so backward shipment will be added.

**Comment by BLR-0118** [comment]
Flow 1 - Delivery
===========
1. /search:
- suggest you use payment.type as "POST-FULFILLMENT" as this is what's enabled for now;
- also applies to other APIs like /init;

2. /init:
- fulfillment.start.contact (for merchant / warehouse) & fulfillment.end.contact (for buyer) should show different contact nos & email IDs;
- if payment.type is ON-ORDER, you also need payment.collected_by;

3. /confirm:
- pls remove bpp_terms, bap_terms as this isn't enabled yet;
- PCC (fulfillment.start.instructions.code & short_desc) are required when order ready_to_ship is "yes";
- DCC may be set only after order is picked up;
- PCC & DCC can't be the same;

4. /update:
- linked order details may be sent only if there is a change in the weight / dim of the goods being shipped (from /confirm);

**Comment by BLR-0118** [comment]
pls resubmit logs only for flow 1
@jagadeeswar577

**Comment by BLR-0118** [comment]
@jagadeeswar577 - ready_to_ship is being sent in /confirm & again in /update; this is incorrect;
clearing logs for v1.2.0 subject to the above being fixed;
